WordPress/gutenberg
# Add a public `bundledPackages`/`excludedExternals` option to DEWP

## What?

This PR allows users of the Dependency Extraction Webpack Plugin "do a simple thing easy" (and complex things still possible) and exclude packages from externalizing in a simple way. It also upstream WooCommerceDEWP feature (https://github.com/woocommerce/woocommerce-admin/pull/7034). Or, in other words, publish the private feature of `BUNDLED_PACKAGES`.

It also includes a tiny clarification in the README (https://github.com/WordPress/gutenberg/pull/45948/commits/e8f3e855c805c08a89d00b26463198e053fb0cbd). The example was using the `entrypoint` for the assets file, whereas the plugin uses `output` config for that. (fixes https://github.com/WordPress/gutenberg/issues/49872)

## Why?

1. To allow doing a simple thing easy, as in the majority of DEWP's use-cases, I've seen `requestToExternal` is used to filter a list of packages.
2. To reduce the confusion of a non-Woo specific (at least conceptually) behavior of providing a simple list of packages to filter, not basing available in WP version of DEWP.
3. To reduce the confusion of `WooCommerceDependencyExtractionWebpackPlugin`'s `bundledPackages` not working for `@wordpress/` packages.

## How?

This PR changes the DEWP constructor to take `excludedExternals` option to use it to filter packages processed in `externalizeWpDeps` before any `requestToExternal` is called.
Turns the old `BUNDLED_PACKAGES` to the option's default value and exposes it as a static property to allow extending it by consumers.

## Testing Instructions




Run automated tests
1. `npx wp-scripts test-unit-js --config test/unit/jest.config.js packages/dependency-extraction-webpack-plugin/`

Manual testing
1. Create a project that uses `@wordpress/components`,  `@wordpress/url`, `@wordpress/icons` and `@wordpress/interface`
2. Setup webpack config to use
```javascript
new DependencyExtractionWebpackPlugin( {
	requestToExternal( package ) { console.log( package ) },
	excludedExternals: [
		...DependencyExtractionWebpackPlugin.excludedExternals,
		'@wordpress/url'
	],
} ),
```
3. Confirm that `@wordpress/url`, `@wordpress/icons` and `@wordpress/interface` gets bundled, and none of them is logged.

## Screenshots or screencast 

## Additional info:
1. I was considering keeping the `bundledPackages` name. This would keep the old name, and make it directly compatible with existing WooCommerce implementation. However:
	1. IMO, it's not the accurate name, as those packages may not get bundled at all. For example, if none of them is actually used & imported.
	2. The behavior is slightly different than in Woo, as the package will not reach `options.requestToExternal`, so this way we avoid breaking change.
	3. We can still have a direct migration path or alias for the old name.
2. I was trying to write tests to check that the default exclusions `/icons` and `/interface` are bundled, but didn't manage for test build to process it. I guess it's related to the monorepo setup. For `lodash` it works as expected.
3. Initially I added it explicitly as 
   ```javascript
	static excludedExternals = // ...
   ```
   but eslint fails to parse it. Probably due to the old ES version used.

----

cc @gziolo as you reviewed my latest DEWP related issues
cc @louwie17 as someone who implemented it in Woo, and could give a context about why it [eventually landed as `bundledPackages` not `excludedExternals`](https://github.com/woocommerce/woocommerce-admin/pull/7034/commits/23c2cd1229a9e000562e97b88267b12d4c342ccd).
